# Quick Introduction to the Workshop

This workshop is copied from The UC Davis Bioinformatics Core, who is giving the same workshop remotely this very same week. I've taken much of what they have put together (all of which based on my prior workshops) and made a few edits specific to how the learning process I believe the students at Makerere will benefit the most.

MOST specifically,

1. This workshop is typically 3 days, where we are stretching it to 5 days.
2. I am including the first day as entirely introductory materials, R and command line.
3. My focus will be on data analysis in R and RStudio and not data reduction. So things will be flipped, first data analysis then data reduction.
4. I am 10 hours behind you, so I will be sleeping when you are in the course.
5. We will use the Github Discussion's page to answer questions.
6. I wills be creating some zoom videos, starting on day 2, Tuesday.
7. We will have live Q&A zoom session starting at day 3, Wednesday at 6am PST, 4pm Uganda time.

## Daily Tasks/Goals

The [page](./schedule) will contain a list of daily tasks to perform as a group. Start the day there.

## Github Discussion Page

Questions and Answers as well as discussions within the group can happen in the Github [Discussions pages](https://github.com/msettles/2022-Uganda-Single-Cell-RNA-Seq-Analysis/discussions).

I've started with an introduction of myself, please comment and tell me about yourselves!

## Workshop Goals

* End to End understanding of the basics of single cell RNAseq experiments
  * Technologies use
  * Experimental design
  * Cost estimation
  * Workflow
  * Data reduction
  * Data analysis

* To work through a complete experiment, starting from raw data to completion (but going backwards), including making a few figures.

## Compute

A portion of this course, data reduction, will need to be conducted on the command line in a Linux environment.
